Cluster size 2 package rbf

instagibbs opened the pull request:
Allows any 2 transaction package with no in-mempool ancestors to do package RBF when directly conflicting with other mempool clusters of size two or less.


Proposed validation steps:
1) If the transaction package is of size 1, legacy rbf rules apply.
2) Otherwise the transaction package consists of a (parent, child) pair with no other in-mempool ancestors (or descendants, obviously), so it is also going to create a cluster of size 2. If larger, fail.
3) The package rbf may not evict more than 100 transactions from the mempool(bip125 rule 5)
4) The package is a single chunk
5) Every directly conflicted mempool transaction is connected to at most 1 other in-mempool transaction (ie the cluster size of the conflict is at most 2).
6) Diagram check: We ensure that the replacement is strictly superior, improving the mempool
7) The total fee of the package, minus the total fee of what is being evicted, is at least the minrelayfee * size of the package (equivalent to bip125 rule 3 and 4)

Post-cluster mempool this will likely be expanded to general package rbf, but this is what we can safely support today.

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +181,36 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+        const CFeeRate replacement_miner_score(replacement_fees, replacement_vsize);
+
+        for (const auto& entry : direct_conflicts) {
+            const bool conflict_is_v3{entry->GetSharedTx()->nVersion == 3};
```
this can't be hit yet; post-v3 it can. I can remove this, or leave a note. It makes some package rbfs more useful

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -228,6 +229,36 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
+    // Tests for CheckMinerScores
+
+    // These tests use modified fees (including prioritisation), not base fees.
+    BOOST_CHECK(CheckMinerScores(entry5_low->GetFee() + entry6_low_prioritised->GetFee() + 1,
```
commit message is old; will fix

instagibbs committed:
CheckPackageMempoolAcceptResult: Check package rbf invariants

instagibbs commented:
After offline discussion with @sdaftuar , I've worked on a set of prospective changes and pushed them to this branch as follow-on commits.

Key change is that we will now only allow package RBF when the conflicted transactions are all in "clusters" of up to size 2. This allows us to calculate mining scores for each conflicted transaction, which means that post-cluster mempool, if we did this right, IIUC, the behavior should match with more general package RBF.

instagibbs commented:
@sdaftuar should be ready for another look

sdaftuar commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +181,54 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+            }
+            if (direct_conflict->GetCountWithAncestors() == 2) {
+                const auto& parent = direct_conflict->GetMemPoolParentsConst().begin();
+                if (parent->get().GetCountWithDescendants() > 1) {
```
Should this be "> 2"?

instagibbs replied:
I have a failing unit test for this :facepalm:

sdaftuar commented:
> size-2 package rbf can have 6 types of conflicts:
> 
> 1. parent conflicts with solo
> 2. parent conflicts with child
> 3. parent conflicts with parent
> 4. child conflicts with parent
> 5. child conflicts with solo
> 6. child conflicts with child

I think if you state it this way, then there are more cases to write out (but they all simplify down to the same thing) -- the parent transaction in the incoming package can directly conflict with {solo tx, child tx, parent tx, both child tx and parent tx}, and so can the child transaction in the package (ie it can also conflict with a solo tx, child tx, parent tx, and both), and you can have both of those going on at the same time, AND we permit the incoming package to conflict with up to 100 such in-mempool transactions that satisfy this "cluster-size-2" restriction.

So the way I'd try to describe this PR is that an incoming package with in-mempool conflicts must satisfy these RBF policy rules:
1) Every direct conflict is connected to at most 1 other in-mempool transaction (ie the cluster size of the conflict is at most 2).
2) The package consists of a (parent, child) pair with no other in-mempool ancestors (or descendants, obviously) (so it is also going to create a cluster of size 2)
3) The package's child tx has greater feerate than the parent tx
4) The total fee of the package, minus the total fee of what is being evicted, is greater than the minrelayfee * size of the package
5) The package feerate (total package fee / total package vsize) must exceed the min(individual feerate, ancestor feerate) of every transaction that would be evicted (direct and indirect conflicts).  For transactions in a cluster of size <= 2, min(feerate, ancestor feerate) exactly captures the mining score of a transaction.

Did I miss anything?

instagibbs commented:
@sdaftuar I took your suggestions, reordered and slightly reworded. Matches my implementation.

sdaftuar reviewed (commented):
I read the code more carefully (other than the tests, which I just skimmed) -- so far this looks good, just had a few comments.  Planning to do some more thorough testing next.

instagibbs commented on src/validation.cpp:
```diff
@@ -1332,11 +1347,8 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
-    std::string err_string;
-    if (txns.size() > 1 && !PackageMempoolChecks(txns, m_total_vsize, package_state)) {
+    // Apply package mempool ancestor/descendant limits.
+    if (!PackageMempoolChecks(txns, m_total_vsize, package_state)) {
```
It's essentially an existing no-op. IIRC `AcceptMultipleTransactions` is only called in *unit tests* with size 1, otherwise it's always > 1.

instagibbs commented on src/validation.cpp:
```diff
@@ -1337,11 +1423,10 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+        // package to spend. Since we already checked conflicts, no transaction can spend a coin
+        // needed by another transaction in the package. We also need to make sure that no package
+        // tx replaces (or replaces the ancestor of) the parent of another package tx. As long as we
+        // check these two things, we don't need to track the coins spent.
```
Pushed a functional test covering this case specifically which is rejected due to existence of mempool ancestors.

I'll try and rephrase soon.

instagibbs replied:
rephrased into the two cases

instagibbs commented:
I responded here: https://delvingbitcoin.org/t/cluster-mempool-rbf-thoughts/156/24?u=instagibbs 

issue 1: awaiting further details on potential diagram check integration vs suggested heuristic I gave
issue 2: known issue, wallet authors shouldn't do that until cluster mempool fixes that\

sdaftuar committed:
[policy] package rbf

Support package RBF where the conflicting package would result
in a mempool cluster of size two, and each of its direct
conflicts are also part of an up-to-size-2 mempool cluster.

This restricted topology allows for exact calculation
of miner scores for each side of the equation, reducing
the surface area for new pins, or incentive-incompatible
replacements.

This allows wallets to create simple CPFP packages
that can fee bump other simple CPFP packages. This,
leveraged with other restrictions such as V3 transactions,
can create pin-resistant applications.

Future package RBF relaxations can be considered when appropriate.

Co-authored-by: glozow <gloriajzhao@gmail.com>
Co-authored-by: Greg Sanders <gsanders87@gmail.com>

glozow committed:
[test] package rbf

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,590 @@
+        self.assert_mempool_contents(expected=package_txns1)
+
+        # Package 2 has a higher feerate but lower absolute fee
+        package_hex2, package_txns2 = self.create_simple_package(coin, parent_fee=DEFAULT_FEE - fee_delta, child_fee=DEFAULT_FEE + fee_delta - Decimal("0.000000001"))
```
too many zeroes on this, it's a 1/10 of a sat

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,590 @@
+        package_hex4, package_txns4 = self.create_simple_package(coin, parent_fee=DEFAULT_FEE, child_fee=DEFAULT_FEE)
+        node.submitpackage(package_hex4)
+        self.assert_mempool_contents(expected=package_txns4, unexpected=[])
+        package_hex5, package_txns5 = self.create_simple_package(coin, parent_fee=DEFAULT_FEE - fee_delta, child_fee=DEFAULT_FEE + fee_delta + Decimal("0.000000001"))
```
too many zeroes on this, it's a 1/10 of a sat

murchandamus commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +181,54 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+        for (const auto& entry : original_transactions) {
+            CFeeRate original_miner_score(entry->GetModFeesWithAncestors(), entry->GetSizeWithAncestors());
+            original_miner_score = std::min(original_miner_score, CFeeRate(entry->GetModifiedFee(), entry->GetTxSize()));
+            if (replacement_miner_score < original_miner_score) {
```
Shouldn’t this be greater or equal to _exceed_ the original?

> 6. The package feerate (total package fee / total package vsize) must exceed the min(individual feerate, ancestor feerate) of every transaction that would be evicted (direct and indirect conflicts).

instagibbs replied:
Probably yeah, tests should catch this case regardless. Will add one after deciding what way to go with heuristic vs diagram check

instagibbs replied:
replaced with diagram check; should be fine now since it must be strictly superior.

murchandamus reviewed (commented):
Concept ACK, did a quick skim review, will review more thoroughly

ismaelsadeeq reviewed (commented):
Concept ACK

instagibbs commented:
Considering an alternative branch with fee diagram checks instead of heuristic here: https://github.com/instagibbs/bitcoin/commits/feefrac_package_rbf

glozow reviewed (commented):
Combed through the code a bit more. Generally approach ACK for "cluster 2 replace cluster 2" idea.

instagibbs commented on src/validation.cpp:
```diff
@@ -1366,7 +1390,7 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
                 std::vector<Wtxid>{ws.m_ptx->GetWitnessHash()};
             results.emplace(ws.m_ptx->GetWitnessHash(),
-                            MempoolAcceptResult::Success(std::move(ws.m_replaced_transactions),
+                            MempoolAcceptResult::Success(std::move(m_replaced_transactions),
```
dealing with duplicates seems strictly superior, will take a look, thanks!

instagibbs replied:
the more work solutions would be to refactor everything to have a "subpackage eval" state, but juice is probably not worth the squeeze until cluster mempool, where "chunk eval" is likely the natural boundary

glozow replied:
yeah I think a temporary `set<Wtxid>` within net processing is probably the easiest

instagibbs replied:
left a comment for future work

instagibbs commented on src/validation.cpp:
```diff
@@ -1067,7 +1072,81 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    // Use the child as the transaction for attributing errors to.
+    const Txid child_hash = child_ws.m_ptx->GetHash();
+    if (const auto err_string{PaysForRBF(m_conflicting_fees, m_total_modified_fees, m_total_vsize,
+                                         m_pool.m_incremental_relay_feerate, child_hash)}) {
```
taken

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +181,54 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
+            }
+        } else {
+            return strprintf("Too many descendants of direct conflict: %zu", desc_count);
+        }
```
cleaner, thanks. will take(minus typos :) )

instagibbs commented on src/policy/rbf.h:
```diff
@@ -106,4 +106,18 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
                                       const uint256& txid);
 
+/** Ensure we are only attempting RBF against connected components we can easily
+ * compute mining scores from.
```
this all got moved internal to mempool

instagibbs commented on src/validation.cpp:
```diff
@@ -507,7 +507,7 @@ class MemPoolAccept
                             /* m_bypass_limits */ false,
```
replaced entirely(I think?)

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -181,3 +181,54 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
     }
     return std::nullopt;
 }
+
```
replaced section of code entirely

instagibbs commented on doc/policy/packages.md:
```diff
@@ -48,8 +48,13 @@ The following rules are enforced for all packages:
      heavily connected, i.e. some transaction in the package is the ancestor or descendant of all
```
changed wording of commit

instagibbs commented on test/functional/mempool_package_onemore.py:
```diff
@@ -55,14 +55,23 @@ def run_test(self):
+        assert_equal(pkg_result["tx-results"][txns[1].getwtxid()]["error"], "bad-txns-inputs-missingorspent")
         # But not if it chains directly off the first transaction
-        replacable_tx = self.wallet.send_self_transfer_multi(from_node=self.nodes[0], utxos_to_spend=[chain[0]])['tx']
+        self.nodes[0].sendrawtransaction(replacable_tx["hex"])
```
fixed

instagibbs commented on src/validation.cpp:
```diff
@@ -972,7 +994,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     }
 
-    m_rbf = !ws.m_conflicts.empty();
+    m_rbf |= !ws.m_conflicts.empty();
```
added

instagibbs commented on test/functional/mempool_package_onemore.py:
```diff
@@ -55,14 +55,23 @@ def run_test(self):
         assert_raises_rpc_error(-26, "too-long-mempool-chain, too many descendants", self.chain_tx, [chain[0], second_chain])
         # ...especially if its > 40k weight
         assert_raises_rpc_error(-26, "too-long-mempool-chain, too many descendants", self.chain_tx, [chain[0]], num_outputs=350)
+        # ...not if it's submitted with other transactions
```
reworded

instagibbs committed:
mempool: Improve logging of replaced transactions

instagibbs commented on src/validation.cpp:
```diff
@@ -1116,19 +1139,18 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
-                FormatMoney(ws.m_modified_fees - ws.m_conflicting_fees),
-                (int)entry->GetTxSize() - (int)ws.m_conflicting_size);
+                FormatMoney(ws.m_modified_fees - m_conflicting_fees),
+                (int)entry->GetTxSize() - (int)m_conflicting_size);
```
done

instagibbs commented:
broke off all the incentive compatibility check stuff into its own PR: https://github.com/bitcoin/bitcoin/pull/29242

Putting this in draft for now to divert attention there

instagibbs committed:
PackageV3Checks: Relax assumptions

Relax assumptions about in-mempool children of in-mempool
parents. With package RBF, we will allow a package of size
2 with conflicts on its parent and reconsider the parent
if its fee is insufficient on its own.

Consider:

TxA (in mempool) <- TxB (in mempool)

TxA (in mempool) <- TxB' (in package, conflicts with TxB) <-
TxC (in package)

If TxB' fails to RBF TxB due to insufficient feerate, the
package TxB' + TxC will be considered. PackageV3Checks
called on TxB' will see an in-mempool parent TxA, and
see the in-mempool child TxB. We cannot assume there is
no in-mempool sibling, rather detect it and fail normally.

Prior to package RBF, this would have failed on the first
conflict in package.

instagibbs commented:
ready for review

glozow reviewed (commented):
I think we had some docs written for doc/policy/packages.md and doc/policy/mempool_replacements.md?

instagibbs commented on src/validation.cpp:
```diff
@@ -1098,16 +1102,15 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
-        // TX_RECONSIDERABLE, because it cannot be bypassed using package validation.
-        // This must be changed if package RBF is enabled.
-        return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY,
+        // Result may change in another package
```
"in a package"?

instagibbs replied:
pushed updated text

glozow commented on src/test/util/txmempool.cpp:
```diff
@@ -68,10 +69,24 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
+            // Replacements can't happen for subpackages larger than 2
+            if (!atmp_result.m_replaced_transactions.value().empty() &&
+                atmp_result.m_wtxids_fee_calculations.has_value() && atmp_result.m_wtxids_fee_calculations.value().size() > 2) {
+                 return strprintf("tx %s was part of a too-large package RBF subpackage",
```
Maybe you can also check the topology/cluster size of this tx in mempool?

instagibbs replied:
added a small check here using `GetEntry` that takes `Wtxid`s.

instagibbs committed:
Fuzz: pass mempool to CheckPackageMempoolAcceptResult

ismaelsadeeq commented on doc/policy/packages.md:
```diff
@@ -48,8 +48,13 @@ The following rules are enforced for all packages:
 
-The following rules are only enforced for packages to be submitted to the mempool (not enforced for
-test accepts):
+* [CPFP Carve Out](./mempool-limits.md#CPFP-Carve-Out) is disabled. (#21800)
```
At first when I see  CPFP carveout disabled I thought we are not longer supporting single transaction CPFP carveout because the linked document is describing rules for a single transaction CPFP carveout.

But from the commit message and reading the code I understand thats not the case, I think we should be explicit that it's package CPFP carvout that is disabled. 

```suggestion
* [Package CPFP Carve Out](./mempool-limits.md#CPFP-Carve-Out) is disabled. (#21800)
```

instagibbs replied:
added an extra phrase, hope that helps?

ismaelsadeeq commented on doc/policy/packages.md:
```diff
@@ -48,8 +48,13 @@ The following rules are enforced for all packages:
+
+   - *Rationale*: This carve out cannot be accurately applied when there are multiple transactions'
+     ancestors and descendants being considered at the same time.
+
```
This also implies that [single conflict package RBF carvout](https://github.com/bitcoin/bitcoin/pull/23711#discussion_r765766768) is not supported, is this worth mentioning?

ismaelsadeeq commented on test/functional/mempool_package_onemore.py:
```diff
@@ -55,14 +55,23 @@ def run_test(self):
         # and the second chain should work just fine
         self.chain_tx([second_chain])
 
         # Make sure we can RBF the chain which used our carve-out rule
```
```suggestion
        # Ensure an individual transaction with single direct conflict can RBF the chain which used our carve-out rule

```

instagibbs replied:
taken

ismaelsadeeq commented on test/functional/mempool_package_onemore.py:
```diff
@@ -55,14 +55,23 @@ def run_test(self):
+        replacement_tx = replaceable_tx["tx"]
+        replacement_tx.vout[0].nValue -= 1000000
+        self.nodes[0].sendrawtransaction(replacement_tx.serialize().hex())
 
```
Maybe test that Package RBF carveout is not supported?
```suggestion

        # But We can not package RBF the chain which used our carveout rule
        replaceable_tx_conflict = self.wallet.create_self_transfer_multi(utxos_to_spend=[chain[0]])
        txns_conflict = [replaceable_tx_conflict["tx"], self.wallet.create_self_transfer_multi(utxos_to_spend=replaceable_tx_conflict["new_utxos"])["tx"]]
        txns_conflict_hex = [tx.serialize().hex() for tx in txns_conflict]
        assert_equal(self.nodes[0].testmempoolaccept(txns_conflict_hex)[0]["reject-reason"], "bip125-replacement-disallowed")
        pkg_result = self.nodes[0].submitpackage(txns_conflict_hex)
        assert "too-long-mempool-chain" in pkg_result["tx-results"][txns_conflict[0].getwtxid()]["error"]
        assert_equal(pkg_result["tx-results"][txns_conflict[1].getwtxid()]["error"], "bad-txns-inputs-missingorspent")
```

instagibbs replied:
would need to do a bit more work to make sure it actually would be attempting a package rbf. I think we can rely on our conflict topology check tests to ensure this, at least for default mempool chains?

ismaelsadeeq replied:
I learned that we can RBF a transaction that gets into the mempool using carveout. I wanted to test that you wouldn't  RBF  that transaction using  a package.

Even if the arrangement attempts to RBF or not, I think the error in the suggestion is how it would fail. We would set `m_allow_carveout` to false while evaluating the first parent transaction and fail its evaluation due to a `TX_MEMPOOL_POLICY` violation of `too-long-mempool-chain`. hence, all the descendants would fail due to missing input.

I just think this is a "nice-to-have" test, but not necessarily required for this PR.

ismaelsadeeq replied:
Ahh I see, it's the same code path.

instagibbs replied:
oh right, neither of these cases will do a package rbf, as the package has in-mempool ancestors.

It would have to be a scenario where the ancestor limits are set to 1, then a package RBF of size 2 replaces that single tx?

ismaelsadeeq replied:
Hmm, I think it will always be the same code path that will be hit because the parent transaction (in the cluster size 2 replacement package) will be detected as conflict with the transaction that was carved out (because it has to go through `PreChecks`), and we don't allow package RBF carveout, so the subpackage evaluation will not be executed?

instagibbs replied:
Right, *even if* the test would manage to hit the case where it would have considered a package rbf(this case won't), it should be caught in both PreChecks and package mempool checks separately.

ismaelsadeeq reviewed (commented):
Continued Reviewing.

- [x] 826ec4c47c0d18a2d9b437b33ea4ceba67e870c1
- [x] 033736bcd9fc16e244e52e72fe7c7ff030690ece
- [x] fa3f94a858
- [x] 8df3b9f7a9
- [x] a6e2372488
- [x] a81d265fb0
- [x] 59e2271167
- [x] ce04c35217
- [x] 2da61cab7e
- [x] ae2fd7b1d6
- [x] d7d697ba40
- [x] 2eb09886f3

ismaelsadeeq commented on doc/policy/packages.md:
```diff
@@ -48,8 +48,13 @@ The following rules are enforced for all packages:
+
+The following rules are only enforced for packages to be submitted to the mempool (not
+enforced for test accepts):
 
```
In `packages.md` there is note stating "Replace By Fee is currently disabled for packages." This is no longer the case, some variants of package can now be RBF'd I think we should indicate that we now support cluster size 2 package replacement into node's mempool stating the new acceptance rules.

instagibbs replied:
added updated comment

instagibbs committed:
doc: update package RBF comment

instagibbs commented on src/validation.cpp:
```diff
@@ -1450,6 +1459,13 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
+        LogPrint(BCLog::MEMPOOL, "replaced %u mempool transactions with 1 new one for %s additional fees, %d delta bytes\n",
+                 m_replaced_transactions.size(),
+                 ws.m_modified_fees - (int)m_conflicting_fees,
+                 ws.m_vsize - (int)m_conflicting_size);
```
done

instagibbs commented on src/validation.cpp:
```diff
@@ -1450,6 +1459,13 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
     }
 
+    if (!m_replaced_transactions.empty()) {
+        LogPrint(BCLog::MEMPOOL, "replaced %u mempool transactions with 1 new one for %s additional fees, %d delta bytes\n",
```
taken

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,604 @@
+        node = self.nodes[0]
+        coin = self.coins.pop()
+
+        # Put simple tx in mempool to chain off of
```
complete sentence, but not grammatically correct or useful, removed

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,604 @@
+        node.submitpackage(package_hex1)
+        self.assert_mempool_contents(expected=package_txns1)
+
+        # Package 2 feerate is below the rate of directly conflicted parent, even though
```
taken

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,604 @@
+        self.assert_mempool_contents(expected=[], unexpected=package_txns1 + package_txns2 + package_txns3)
+        self.generate(node, 1)
+
+    def test_too_numerous_pkg(self):
```
taken

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,604 @@
+        self.assert_mempool_contents(expected=package_txns2, unexpected=package_txns1)
+        self.generate(node, 1)
+
+    def test_too_numerous_ancestors(self):
```
it's a package rbf test, I think name is fine

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,604 @@
+        self.assert_mempool_contents(expected=package_txns3, unexpected=package_txns2)
+
+    def test_package_rbf_partial(self):
+        self.log.info("Test that package RBF works when a transaction was already submitted")
```
Hmm, yeah I don't think this test is meaningful. Removed.

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,604 @@
+        assert_equal(f"package RBF failed: too many potential replacements, rejecting replacement {package_parent['tx'].rehash()}; too many potential replacements (125 > 100)\n", pkg_results["package_msg"])
+        self.generate(node, 1)
+
+    def test_package_rbf_conflicting_conflicts(self):
```
done

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,604 @@
+            "-maxmempool=5",
+        ]]
+
+    def fill_mempool(self):
```
forgot to remove, done

instagibbs commented on src/test/txpackage_tests.cpp:
```diff
@@ -842,4 +843,138 @@ BOOST_FIXTURE_TEST_CASE(package_cpfp_tests, TestChain100Setup)
+        Package package2{tx_parent_2, tx_child_2};
+        // 1 parent paying 200sat, 1 child paying 1300. Both v2.
+        Package package3{tx_parent_3, tx_child_3};
+        // In all packages, the parents conflict with each other
```
fixed

instagibbs commented on src/test/txpackage_tests.cpp:
```diff
@@ -842,4 +843,138 @@ BOOST_FIXTURE_TEST_CASE(package_cpfp_tests, TestChain100Setup)
+            tx_parent_2, /*input_vout=*/0, /*input_height=*/101,
+            child_key, child_spk, coinbase_value - 800 - 200, /*submit=*/false));
+
+        CTransactionRef tx_parent_3 = MakeTransactionRef(CreateValidMempoolTransaction(
```
conceptually the test is cleaner if each parent is different, so I updated the parent to be slightly different, and asserted the parents are all have different txids

instagibbs commented on src/validation.cpp:
```diff
@@ -1123,7 +1126,74 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    }
+
+
+    // Calculate all conflicting entries and enforce Rules 2 and 5.
```
Deleting this comment as I don't think it's helpful

instagibbs commented:
all feedback should be addressed now

ismaelsadeeq reviewed (approved):
Code Review ACK 7cede4caa0a6b3dd57397d96ee98239fb890ca32

My comments were addressed and I have reviewed all the commits and test this locally on regtest.

I've fuzz this PR for a while locally and no any crash.

instagibbs commented:
rebased on master due to conflict from https://github.com/bitcoin/bitcoin/pull/29906

ismaelsadeeq reviewed (commented):
reACK 69deec6fd074e8524dd11103834739b02f50e814

instagibbs commented on src/validation.cpp:
```diff
@@ -1126,7 +1129,72 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+
+    // We're in package RBF context; replacement proposal must be size 2
+    if (workspaces.size() != 2) {
+        return package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package RBF failed: replacing cluster not size two");
```
Should all these failures be `PCKG_POLICY`? We have individual errors which are reported as well, and now with 1P1C relay, we probably want to act on failures at p2p layer?

instagibbs replied:
update: in master we're now processing these `PCKG_POLICY` cases, so should be ok? @glozow

instagibbs replied:
IIUC It will leave the "individual failures" in place which should be:
parent: TX_RECONSIDERABLE
child: TX_MISSING_INPUTS

instagibbs commented:
rebased to pick up 1P1C relay

instagibbs commented on src/validation.cpp:
```diff
@@ -1146,6 +1146,12 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
     }
 
+    // Should be parent->child topology
+    const auto parent_result{FindInPackageParents(txns, txns.back())};
```
Oh even easier than what I replaced it with, duh

instagibbs replied:
pushed with the `Assume()` check inline with the "making sure package is cluster size two" check

instagibbs commented:
rebased due to conflict on master

glozow commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,9 @@ The following rules are enforced for all packages:
+* Only limited package replacements are currently considered. (#28984)
 
-   - Package RBF may be enabled in the future.
+   - More general package RBF may be enabled in the future.
```
Could add more detailed docs for what the rules are in package RBF?

- all to-be-replaced signal (bip125 or v3)
- package is 1p1c, only in clusters up to size 2
- to-be-replaced all in clusters up to size 2
- no more than 100 total to-be-replaced
- total fees of package > all fees being replaced, at incremental relay feerate
- must improve feerate diagram
- parent feerate must be < package feerate

instagibbs replied:
@glozow pushed some more detailed docs, thanks!

instagibbs commented on src/validation.cpp:
```diff
@@ -1126,7 +1125,71 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+                                     "package RBF failed: parent paying for child anti-DoS", "");
+    }
+
+    // Check if it's economically rational to mine this package rather than the ones it replaces.
+    if (const auto err_tup{ImprovesFeerateDiagram(m_pool, direct_conflict_iters, m_all_conflicts, m_total_modified_fees, m_total_vsize)}) {
+        return package_state.Invalid(PackageValidationResult::PCKG_POLICY,
+                                     "package RBF failed: " + err_tup.value().second, "");
+    }
```
> is a superset of the rules

PaysForRBF has incremental rate on top, so `ImprovesFeerateDiagram` is not a strict superset fwiw.

That said, I don't think swapping the order is a problem and would indeed be cheaper.

glozow replied:
True, it's only a superset of Rule 3 👍

instagibbs replied:
swapped them and updated tests accordingly

instagibbs commented on src/validation.cpp:
```diff
@@ -1087,7 +1125,76 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+
+    // We're in package RBF context; replacement proposal must be size 2
+    if (workspaces.size() != 2 || !Assume(IsChildWithParents(txns))) {
+        return package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package RBF failed: replacing cluster not size two");
```
done

instagibbs commented on src/validation.cpp:
```diff
@@ -1126,7 +1125,71 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+                                     "package RBF failed: parent paying for child anti-DoS", "");
+    }
+
+    // Check if it's economically rational to mine this package rather than the ones it replaces.
```
done

instagibbs commented on src/validation.cpp:
```diff
@@ -1126,7 +1125,71 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    // since it would result in a cluster larger than 2
+    for (const auto& ws : workspaces) {
+        if (!ws.m_ancestors.empty()) {
+            return package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package RBF failed: replacing cluster with ancestors not size two");
```
done

instagibbs commented on src/validation.cpp:
```diff
@@ -1126,7 +1125,71 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+            CFeeRate(parent_ws.m_modified_fees + child_ws.m_modified_fees, parent_ws.m_vsize + child_ws.m_vsize)) {
+        return package_state.Invalid(PackageValidationResult::PCKG_POLICY,
+                                     "package RBF failed: parent paying for child anti-DoS", "");
+    }
```
done

instagibbs commented:
@glozow thanks for the review, all comments addressed

instagibbs commented:
silent merge conflict with https://github.com/bitcoin/bitcoin/pull/29939

glozow commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,595 @@
+        package_hex5, package_txns5 = self.create_simple_package(coin, parent_fee=DEFAULT_FEE, child_fee=DEFAULT_FEE)
+        node.submitpackage(package_hex5)
+        self.assert_mempool_contents(expected=package_txns5)
+        package_hex6, package_txns6 = self.create_simple_package(coin, parent_fee=DEFAULT_FEE + fee_delta, child_fee=DEFAULT_FEE - (fee_delta / 2))
```
Given that `fee_delta` is a multiple of `DEFAULT_FEE` and you're using `fee_delta / 2` here... I think the general readability of this test would be improved if you created a constant base unit `FEE` which was equal to, say, `DEFAULT_FEE / 10` (or `104 * 10` which would give most of the transactions whole number feerates), and then used `FEE`, `2*FEE`, `8*FEE`, etc. in all the tests.

instagibbs replied:
I created `DEFAULT_CHILD_FEE` and used that pretty much everywhere it made sense

glozow commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,595 @@
+        self.generate(node, 1)
+
+    def test_child_conflicts_parent_mempool_ancestor(self):
+        fill_mempool(self, self.nodes[0], self.wallet)
```
with rebase
```suggestion
        fill_mempool(self, self.nodes[0])
```

instagibbs replied:
done

glozow commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,27 @@ The following rules are enforced for all packages:
+
+   - *Rationale*: Basic support for package RBF can be used by wallets
+     by making chains of no longer than two, then directly conflicting
+     those chains when needed. Combined with V3 transactions this can
```
Maybe there should be some v3 test coverage (including nice 0fee parent package RBF)?
```
diff --git a/test/functional/mempool_accept_v3.py b/test/functional/mempool_accept_v3.py
index 8285b82c19..990d36359e 100755
--- a/test/functional/mempool_accept_v3.py
+++ b/test/functional/mempool_accept_v3.py
@@ -579,6 +579,32 @@ class MempoolAcceptV3(BitcoinTestFramework):
         )
         self.check_mempool([tx_with_multi_children["txid"], tx_with_sibling3_rbf["txid"], tx_with_sibling2["txid"]])
 
+    @cleanup(extra_args=["-acceptnonstdtxn=1"])
+    def test_package_rbf(self):
+        self.log.info("Test package RBF: v3 0-fee parent + high-fee child replaces parent's conflicts")
+        node = self.nodes[0]
+        # Reuse the same coins so that the transactions conflict with one another.
+        parent_coin = self.wallet.get_utxo(confirmed_only=True)
+
+        # package1 pays default fee on both transactions
+        parent1 = self.wallet.create_self_transfer(utxo_to_spend=parent_coin, version=3)
+        child1 = self.wallet.create_self_transfer(utxo_to_spend=parent1["new_utxo"], version=3)
+        package_hex1 = [parent1["hex"], child1["hex"]]
+        fees_package1 = parent1["fee"] + child1["fee"]
+        submitres1 = node.submitpackage(package_hex1)
+        assert_equal(submitres1["package_msg"], "success")
+        self.check_mempool([parent1["txid"], child1["txid"]])
+
+        # package2 has a 0-fee parent (conflicting with package1) and very high fee child
+        parent2 = self.wallet.create_self_transfer(utxo_to_spend=parent_coin, fee=0, fee_rate=0, version=3)
+        child2 = self.wallet.create_self_transfer(utxo_to_spend=parent2["new_utxo"], fee=fees_package1*10, version=3)
+        package_hex2 = [parent2["hex"], child2["hex"]]
+
+        submitres2 = node.submitpackage(package_hex2)
+        assert_equal(submitres2["package_msg"], "success")
+        assert_equal(set(submitres2["replaced-transactions"]), set([parent1["txid"], child1["txid"]]))
+        self.check_mempool([parent2["txid"], child2["txid"]])
+
 
     def run_test(self):
         self.log.info("Generate blocks to create UTXOs")
@@ -598,6 +624,7 @@ class MempoolAcceptV3(BitcoinTestFramework):
         self.test_reorg_2child_rbf()
         self.test_v3_sibling_eviction()
         self.test_reorg_sibling_eviction_1p2c()
+        self.test_package_rbf()
 
 
 if __name__ == "__main__":
```

instagibbs replied:
was wondering about this case, added with some modifications for test simplicity

instagibbs replied:
done in follow-up, I had removed V3 in test, not the actual doc

glozow reviewed (commented):
some test review

instagibbs commented on src/validation.cpp:
```diff
@@ -1087,7 +1125,80 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    if (parent_feerate >= package_feerate) {
+        return package_state.Invalid(PackageValidationResult::PCKG_POLICY,
+                                     "package RBF failed: parent paying for child replacement",
+                                     strprintf("package feerate is %s while parent feerate is %s", package_feerate.ToString(), parent_feerate.ToString()));
```
taken

instagibbs commented on src/validation.cpp:
```diff
@@ -1087,7 +1125,80 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    const auto& child_ws = workspaces[1];
+
+    // Use the child as the transaction for attributing errors to.
+    const Txid child_hash = child_ws.m_ptx->GetHash();
```
done

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,595 @@
+    assert_greater_than_or_equal,
+    assert_raises_rpc_error,
+    assert_equal,
+    fill_mempool,
```
done

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,595 @@
+        self.assert_mempool_contents(expected=expected_txns)
+
+        # Now make conflicting packages for each coin
+        package_hex1, package_txns1 = self.create_simple_package(coin1, DEFAULT_FEE, DEFAULT_FEE * 4)
```
picked a constant and used it

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,595 @@
+        self.ctr += 1
+        parent_result1 = self.wallet.create_self_transfer(
+            fee_rate=0,
+            fee=DEFAULT_FEE,
```
removed a series of these ignored args

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,595 @@
+        node.submitpackage(package_hex1)
+        self.assert_mempool_contents(expected=package_txns1)
+        # The first two transactions have the same conflicts
+        package_duplicate_conflicts_hex = [package_hex2[0]] + package_hex3
```
Agreed, it's perhaps an older badly done version of #30066, removed

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,595 @@
+        # Finally, evict MAX_REPLACEMENT_CANDIDATES
+        package_parent = self.wallet.create_self_transfer_multi(utxos_to_spend=parent_coins[:-1], fee_per_output=10000)
+        package_child = self.wallet.create_self_transfer(fee_rate=10000*DEFAULT_FEE, utxo_to_spend=package_parent["new_utxos"][0])
+        pkg_results = node.submitpackage([package_parent["hex"], package_child["hex"]], maxfeerate=0)
```
I turned off maxfeerate, jacked up the child feerates, and made constants for both

instagibbs commented:
rebased due to conflict

t-bast commented:
While I wouldn't trust myself to correctly code-review this PR, I'd be happy to work on e2e tests that would leverage this for lightning channels fee-bumping (based on [eclair](https://github.com/acinq/eclair)) if it can help validate the logic and get this PR merged.

I'd like to highlight again how important this feature is for lightning (and probably for many other L2 protocols on top of bitcoin today). This is the critical step that allows us to mitigate pinning of a commitment transaction, and guarantee that we're able to set the fees of our commitment package to a value that should ensure timely confirmation.

instagibbs commented:
sorry forgot to link https://github.com/bitcoin/bitcoin/pull/30072 here, added to OP

instagibbs commented:
rebased on latest #30072

instagibbs commented:
rebased on master :rocket:

instagibbs commented on src/policy/v3_policy.cpp:
```diff
@@ -135,10 +134,7 @@ std::optional<std::string> PackageV3Checks(const CTransactionRef& ptx, int64_t v
-            // catches mempool siblings and sibling eviction is not extended to packages. Also, if the package consists of connected transactions,
-            // any tx having a mempool ancestor would mean the package exceeds ancestor limits.
-            if (!Assume(!parent_info.m_has_mempool_descendant)) {
+            if (parent_info.m_has_mempool_descendant) {
```
> is this commit (848c4e5, "PackageV3Checks: Relax assumptions") strictly needed for this PR or already preparation for future work?

If it's not removed the next commit will cause failures.

It hits pretty in the fuzzing `package_eval` fuzzing target. I could add a unit test for it, but it doesn't really test much, except that we don't crash right before removing the soon to be erroneous assertion. Here's an input you can run to verify: 
```
EQQAAAgAAAAAAAAAAAAAAAUAAAAAAAAAAAAAAAA6AAAACAAADQAAAAAAAAAAAAAAAAAAaQFVKGEA
awAAAAAABQAAAAAAAAAAAAAAAAA6AAAACAAADQAAAAAAAAAAAAAAAAAAaQFVKGF0YWNhcnJpZXJz
aXplVVUAAAEAAAAAAFURVQEuaQEA/wAAAigBbwAAOzsBswBiaQFpAWR0YWNhcnJpZXJzaXplVVUA
AAAAAAAAAFURVQFkaQEA/wAAAigBbwAAOzsBswBiaQFVZCgFaWkBAQA=
```

> isn't possible due to V3 ancestor limits.

It's only impossible due to `PackageV3Checks`, since `TxC` has no in-mempool ancestors at `PreChecks/SingleV3Checks` time.

ismaelsadeeq replied:
> It hits in the fuzzing package_eval fuzzing target.

I was able to reproduce this locally.

glozow commented on src/test/txpackage_tests.cpp:
```diff
@@ -938,4 +939,140 @@ BOOST_FIXTURE_TEST_CASE(package_cpfp_tests, TestChain100Setup)
+
+        LOCK(m_node.mempool->cs);
+        const auto submit1 = ProcessNewPackage(m_node.chainman->ActiveChainstate(), *m_node.mempool, package1, /*test_accept=*/false, std::nullopt);
+        BOOST_CHECK_MESSAGE(submit1.m_state.IsValid(), "Package validation unexpectedly failed: " << submit1.m_state.GetRejectReason());
+        auto it_parent_1 = submit1.m_tx_results.find(tx_parent->GetWitnessHash());
+        auto it_child_1 = submit1.m_tx_results.find(tx_child_1->GetWitnessHash());
+        BOOST_CHECK(it_parent_1 != submit1.m_tx_results.end());
+        BOOST_CHECK(it_child_1 != submit1.m_tx_results.end());
+        BOOST_CHECK_EQUAL(it_parent_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
+        BOOST_CHECK_EQUAL(it_child_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
+        expected_pool_size += 2;
+        BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
+        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_parent->GetHash())));
+        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_child_1->GetHash())));
```
Should use `CheckPackageMempoolAcceptResult` to check that results are populated as expected, which also means you can remove a few of these checks:

```suggestion
        if (auto err_1{CheckPackageMempoolAcceptResult(package1, submit1, /*expect_valid=*/true, m_node.mempool.get())}) {
            BOOST_ERROR(err_1.value());
        }
        auto it_parent_1 = submit1.m_tx_results.find(tx_parent->GetWitnessHash());
        auto it_child_1 = submit1.m_tx_results.find(tx_child_1->GetWitnessHash());

        BOOST_CHECK_EQUAL(it_parent_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
        BOOST_CHECK_EQUAL(it_child_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
        expected_pool_size += 2;
        BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
```

instagibbs replied:
great, added a little commentary on top and repeated this for 2nd package

glozow commented on src/test/txpackage_tests.cpp:
```diff
@@ -938,4 +939,140 @@ BOOST_FIXTURE_TEST_CASE(package_cpfp_tests, TestChain100Setup)
+        BOOST_CHECK_EQUAL(it_child_3->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
+        BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
+        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_parent_3->GetHash())));
+        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_child_3->GetHash())));
```
Can check `m_wtxids_fee_calculations` and `m_effective_feerate`.

Wrote up the changes for this + using `CheckPackageMempoolAcceptResult`:
```diff --git a/src/test/txpackage_tests.cpp b/src/test/txpackage_tests.cpp
index cd69fce2f5..b68686cfcb 100644
--- a/src/test/txpackage_tests.cpp
+++ b/src/test/txpackage_tests.cpp
@@ -975,28 +975,22 @@ BOOST_FIXTURE_TEST_CASE(package_rbf_tests, TestChain100Setup)
 
         LOCK(m_node.mempool->cs);
         const auto submit1 = ProcessNewPackage(m_node.chainman->ActiveChainstate(), *m_node.mempool, package1, /*test_accept=*/false, std::nullopt);
-        BOOST_CHECK_MESSAGE(submit1.m_state.IsValid(), "Package validation unexpectedly failed: " << submit1.m_state.GetRejectReason());
-        auto it_parent_1 = submit1.m_tx_results.find(tx_parent->GetWitnessHash());
-        auto it_child_1 = submit1.m_tx_results.find(tx_child_1->GetWitnessHash());
-        BOOST_CHECK(it_parent_1 != submit1.m_tx_results.end());
-        BOOST_CHECK(it_child_1 != submit1.m_tx_results.end());
-        BOOST_CHECK_EQUAL(it_parent_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
-        BOOST_CHECK_EQUAL(it_child_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
+        if (auto err_1{CheckPackageMempoolAcceptResult(package1, submit1, /*expect_valid=*/true, m_node.mempool.get())}) {
+            BOOST_ERROR(err_1.value());
+        }
         expected_pool_size += 2;
-        BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_parent->GetHash())));
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_child_1->GetHash())));
 
         const auto submit2 = ProcessNewPackage(m_node.chainman->ActiveChainstate(), *m_node.mempool, package2, /*test_accept=*/false, std::nullopt);
+        if (auto err_2{CheckPackageMempoolAcceptResult(package2, submit2, /*expect_valid=*/true, m_node.mempool.get())}) {
+            BOOST_ERROR(err_2.value());
+        }
         auto it_parent_2 = submit2.m_tx_results.find(tx_parent->GetWitnessHash());
         auto it_child_2 = submit2.m_tx_results.find(tx_child_2->GetWitnessHash());
-        BOOST_CHECK(it_parent_2 != submit2.m_tx_results.end());
-        BOOST_CHECK(it_child_2 != submit2.m_tx_results.end());
-        BOOST_CHECK_MESSAGE(submit2.m_state.IsValid(), "Package validation unexpectedly failed" << submit2.m_state.GetRejectReason());
         BOOST_CHECK_EQUAL(it_parent_2->second.m_result_type, MempoolAcceptResult::ResultType::MEMPOOL_ENTRY);
         BOOST_CHECK_EQUAL(it_child_2->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
         BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_child_2->GetHash())));
+
+        // child1 has been replaced
         BOOST_CHECK(!m_node.mempool->exists(GenTxid::Txid(tx_child_1->GetHash())));
     }
 
@@ -1034,44 +1028,60 @@ BOOST_FIXTURE_TEST_CASE(package_rbf_tests, TestChain100Setup)
         Package package3{tx_parent_3, tx_child_3};
 
         const auto submit1 = ProcessNewPackage(m_node.chainman->ActiveChainstate(), *m_node.mempool, package1, false, std::nullopt);
-        BOOST_CHECK_MESSAGE(submit1.m_state.IsValid(), "Package validation unexpectedly failed" << submit1.m_state.GetRejectReason());
+        if (auto err_1{CheckPackageMempoolAcceptResult(package1, submit1, /*expect_valid=*/true, m_node.mempool.get())}) {
+            BOOST_ERROR(err_1.value());
+        }
         auto it_parent_1 = submit1.m_tx_results.find(tx_parent_1->GetWitnessHash());
         auto it_child_1 = submit1.m_tx_results.find(tx_child_1->GetWitnessHash());
-        BOOST_CHECK(it_parent_1 != submit1.m_tx_results.end());
-        BOOST_CHECK(it_child_1 != submit1.m_tx_results.end());
         BOOST_CHECK_EQUAL(it_parent_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
         BOOST_CHECK_EQUAL(it_child_1->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
+        BOOST_CHECK_EQUAL(it_parent_1->second.m_effective_feerate.value().GetFee(GetVirtualTransactionSize(*tx_parent_1)), 200);
+        BOOST_CHECK_EQUAL(it_child_1->second.m_effective_feerate.value().GetFee(GetVirtualTransactionSize(*tx_child_1)), 200);
+        BOOST_CHECK(it_parent_1->second.m_replaced_transactions.empty());
+        BOOST_CHECK(it_child_1->second.m_replaced_transactions.empty());
         expected_pool_size += 2;
         BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_parent_1->GetHash())));
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_child_1->GetHash())));
 
         // This replacement is actually not package rbf; the parent carries enough fees
         // to replace the entire package on its own.
         const auto submit2 = ProcessNewPackage(m_node.chainman->ActiveChainstate(), *m_node.mempool, package2, false, std::nullopt);
-        BOOST_CHECK_MESSAGE(submit2.m_state.IsValid(), "Package validation unexpectedly failed" << submit2.m_state.GetRejectReason());
+        if (auto err_2{CheckPackageMempoolAcceptResult(package2, submit2, /*expect_valid=*/true, m_node.mempool.get())}) {
+            BOOST_ERROR(err_2.value());
+        }
         auto it_parent_2 = submit2.m_tx_results.find(tx_parent_2->GetWitnessHash());
         auto it_child_2 = submit2.m_tx_results.find(tx_child_2->GetWitnessHash());
-        BOOST_CHECK(it_parent_2 != submit2.m_tx_results.end());
-        BOOST_CHECK(it_child_2 != submit2.m_tx_results.end());
         BOOST_CHECK_EQUAL(it_parent_2->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
         BOOST_CHECK_EQUAL(it_child_2->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
+        BOOST_CHECK_EQUAL(it_parent_2->second.m_effective_feerate.value().GetFee(GetVirtualTransactionSize(*tx_parent_2)), 800);
+        BOOST_CHECK_EQUAL(it_child_2->second.m_effective_feerate.value().GetFee(GetVirtualTransactionSize(*tx_child_2)), 200);
+
+        // parent2 replaced both transactions
+        BOOST_CHECK(it_parent_2->second.m_replaced_transactions.size() == 2);
+        BOOST_CHECK(it_child_2->second.m_replaced_transactions.empty());
         BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_parent_2->GetHash())));
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_child_2->GetHash())));
 
         // Package RBF, in which the replacement transaction's child sponsors the fees to meet RBF feerate rules
         const auto submit3 = ProcessNewPackage(m_node.chainman->ActiveChainstate(), *m_node.mempool, package3, false, std::nullopt);
-        BOOST_CHECK_MESSAGE(submit3.m_state.IsValid(), "Package validation unexpectedly failed" << submit3.m_state.GetRejectReason());
+        if (auto err_3{CheckPackageMempoolAcceptResult(package3, submit3, /*expect_valid=*/true, m_node.mempool.get())}) {
+            BOOST_ERROR(err_3.value());
+        }
         auto it_parent_3 = submit3.m_tx_results.find(tx_parent_3->GetWitnessHash());
         auto it_child_3 = submit3.m_tx_results.find(tx_child_3->GetWitnessHash());
-        BOOST_CHECK(it_parent_3 != submit3.m_tx_results.end());
-        BOOST_CHECK(it_child_3 != submit3.m_tx_results.end());
         BOOST_CHECK_EQUAL(it_parent_3->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
         BOOST_CHECK_EQUAL(it_child_3->second.m_result_type, MempoolAcceptResult::ResultType::VALID);
+
+        // package3 was considered as a package to replace both package2 transactions
+        BOOST_CHECK(it_parent_3->second.m_replaced_transactions.size() == 2);
+        BOOST_CHECK(it_child_3->second.m_replaced_transactions.empty());
+
+        std::vector<Wtxid> expected_package3_wtxids({tx_parent_3->GetWitnessHash(), tx_child_3->GetWitnessHash()});
+        const auto package3_total_vsize{GetVirtualTransactionSize(*tx_parent_3) + GetVirtualTransactionSize(*tx_child_3)};
+        BOOST_CHECK(it_parent_3->second.m_wtxids_fee_calculations.value() == expected_package3_wtxids);
+        BOOST_CHECK(it_child_3->second.m_wtxids_fee_calculations.value() == expected_package3_wtxids);
+        BOOST_CHECK_EQUAL(it_parent_3->second.m_effective_feerate.value().GetFee(package3_total_vsize), 1499);
+        BOOST_CHECK_EQUAL(it_child_3->second.m_effective_feerate.value().GetFee(package3_total_vsize), 1499);
+
         BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_parent_3->GetHash())));
-        BOOST_CHECK(m_node.mempool->exists(GenTxid::Txid(tx_child_3->GetHash())));
     }
 
 }

```

instagibbs replied:
taken with only light modifications

glozow commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,590 @@
+        self.extra_args = [[
+            "-datacarriersize=100000",
+            "-maxmempool=5",
+            "-acceptnonstdtxn=1", # TRUC test; remove once standard
```
Needs rebase for #29496?

instagibbs replied:
yep, rebased + removed

glozow commented on src/validation.cpp:
```diff
@@ -1391,6 +1393,13 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
+    if (!m_subpackage.m_replaced_transactions.empty()) {
+        LogPrint(BCLog::MEMPOOL, "replaced %u mempool transactions with %u new one(s) for %s additional fees, %d delta bytes\n",
+                 m_subpackage.m_replaced_transactions.size(), workspaces.size(),
+                 m_subpackage.m_total_modified_fees - static_cast<int>(m_subpackage.m_conflicting_fees),
+                 m_subpackage.m_total_vsize - static_cast<int>(m_subpackage.m_conflicting_size));
```
aa3a4da248fe9d4a765fa4c80797ca333ee37f07: is the static cast necessary?

instagibbs replied:
removed the static cast for `m_conflicting_fees` since that's the same type as `m_total_modified_fees`, but left the other static casts in place since explicit is better than implicit

glozow commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -307,7 +307,7 @@ FUZZ_TARGET(tx_package_eval, .init = initialize_tx_pool)
             // we can still verify that the contents of m_tx_results are consistent with m_state.
             const bool expect_valid{result_package.m_state.IsValid()};
-            Assert(!CheckPackageMempoolAcceptResult(txs, result_package, expect_valid, nullptr));
+            Assert(!CheckPackageMempoolAcceptResult(txs, result_package, expect_valid, &tx_pool));
```
5dd74f9c6ddc218417bda841f493ca2ed3f2b327

I don't 100% remember why we did this, but I think it was so we don't check mempool contents when `test_accept=true`?

instagibbs replied:
This should only be called when not `test_accept`, right?

glozow replied:
I think so...

glozow replied:
Aha yes. Can resolve.

glozow commented on src/test/util/txmempool.cpp:
```diff
@@ -82,6 +82,19 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
+            if (atmp_result.m_wtxids_fee_calculations.has_value() && atmp_result.m_wtxids_fee_calculations.value().size() == 2) {
+                for (const auto& fee_calc_wtxid : atmp_result.m_wtxids_fee_calculations.value()) {
+                    const auto& entry = *Assert(mempool->GetEntry(fee_calc_wtxid));
+                    if (entry.GetCountWithDescendants() + entry.GetCountWithAncestors() != 3) {
+                        return strprintf("tx %s has too many ancestors or descendants for a package rbf", fee_calc_wtxid.ToString());
+                    }
```
91a7e8ba7606b68401b29d2609c240a604ab6fe1: this isn't a perfect test for cluster size 2... why not call `CheckConflictTopology`?

instagibbs replied:
Directly checked cluster size instead.

glozow replied:
Nice, agree that's better. However e4e14fedc9622ca7cfc40af4f2aa70ed4bb7daa6 is now unnecessary and can be dropped.

instagibbs replied:
even better. dropped.

glozow commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,27 @@ The following rules are enforced for all packages:
+
+   - No more than MAX_REPLACEMENT_CANDIDATES transactions can be replaced.
+
+   - Total fee plus incremental relay fee must be paid (ala bip125 rules 3 and 4).
```
No mentions of BIP125 outside of signaling pls
```suggestion
   - Replacements must pay more total total fees at the incremental relay fee (analogous to regular [replacement rules](./mempool-replacements.md) 3 and 4).
```

instagibbs replied:
sure, taken

glozow reviewed (commented):
looks pretty good 7b74cbf6ed7cafdbd458471ae89fa88c52256278

instagibbs commented on src/test/txpackage_tests.cpp:
```diff
@@ -938,4 +939,140 @@ BOOST_FIXTURE_TEST_CASE(package_cpfp_tests, TestChain100Setup)
+        auto mtx_parent = CreateValidMempoolTransaction(/*input_transaction=*/m_coinbase_txns[0], /*input_vout=*/0,
+                                                        /*input_height=*/0, /*input_signing_key=*/coinbaseKey,
+                                                        /*output_destination=*/parent_spk,
+                                                        /*output_amount=*/coinbase_value - 200, /*submit=*/false);
```
adapted more of the test to use this value

instagibbs commented on src/validation.cpp:
```diff
@@ -1141,7 +1140,80 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    const auto& parent_ws = workspaces[0];
+    const auto& child_ws = workspaces[1];
+
+    // The aggregated set of conflicts cannot exceed 100.
```
taken

ismaelsadeeq commented on src/validation.cpp:
```diff
@@ -524,7 +524,7 @@ class MemPoolAccept
                             /* m_bypass_limits */ false,
```
In d3d653dce34ad1f3235869b8a54f9572e89e5b68  [policy] package rbf 

Update places where it is claimed that package RBF is disabled?


```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 8372fcadca1..d18fc1c0f2c 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -602,8 +602,8 @@ public:
     /**
      * Submission of a subpackage.
      * If subpackage size == 1, calls AcceptSingleTransaction() with adjusted ATMPArgs to avoid
-     * package policy restrictions like no CPFP carve out (PackageMempoolChecks) and disabled RBF
-     * (m_allow_replacement), and creates a PackageMempoolAcceptResult wrapping the result.
+     * package policy restrictions like no CPFP carve out (PackageMempoolChecks)
+     * and creates a PackageMempoolAcceptResult wrapping the result.
      *
      * If subpackage size > 1, calls AcceptMultipleTransactions() with the provided ATMPArgs.
      *
@@ -950,7 +950,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     ws.m_iters_conflicting = m_pool.GetIterSet(ws.m_conflicts);
 
     // Note that these modifications are only applicable to single transaction scenarios;
-    // carve-outs and package RBF are disabled for multi-transaction evaluations.
+    // carve-outs are disabled for multi-transaction evaluations.
     CTxMemPool::Limits maybe_rbf_limits = m_pool.m_opts.limits;
 
     // Calculate in-mempool ancestors, up to a limit.
```

instagibbs replied:
done, and found one more place with out of date comment

ismaelsadeeq commented on src/test/txpackage_tests.cpp:
```diff
@@ -938,4 +939,149 @@ BOOST_FIXTURE_TEST_CASE(package_cpfp_tests, TestChain100Setup)
+        Package package1{tx_parent_1, tx_child_1};
+        // 1 parent paying 800sat, 1 child paying 200sat.
+        Package package2{tx_parent_2, tx_child_2};
+        // 1 parent paying 199sat, 1 child paying 1300.
```
nit:
```suggestion
        // 1 parent paying 199sat, 1 child paying 1300sat.
```

instagibbs replied:
done

glozow reviewed (commented):
ACK dd364f9a813221956ec0b19f45db231b148491c2. non-blocking nits that can go in a followup. Also needs a release note.

instagibbs commented on src/validation.cpp:
```diff
@@ -1434,11 +1530,14 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
-        // updated if package replace-by-fee is allowed in the future.
-        assert(!args.m_allow_replacement);
-        assert(!m_subpackage.m_rbf);
+        // package to spend. If there are no conflicts, no transaction can spend a coin
+        // needed by another transaction in the package. We also need to make sure that no package
+        // tx replaces (or replaces the ancestor of) the parent of another package tx. As long as we
+        // check these two things, we don't need to track the coins spent.
+        // If there are conflicts, PackageMempoolChecks() ensures later that any package RBF attempt
```
taken

instagibbs commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,28 @@ The following rules are enforced for all packages:
+
+   - Packages are 1-parent-1-child, with no in-mempool ancestors of the package.
+
+   - All conflicting clusters must be clusters of up to size 2.
```
defined it inline

instagibbs commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,28 @@ The following rules are enforced for all packages:
+
+   - Parent feerate must be lower than package feerate.
+
+   - Must improve feerate diagram. (#29242)
```
Added delving link but for Future Work I'm wondering longer term where cluster mempool design stuff should live.

instagibbs commented on src/test/util/txmempool.cpp:
```diff
@@ -82,6 +82,15 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
                                 wtxid.ToString());
         }
 
+        if (!atmp_result.m_replaced_transactions.empty() && mempool) {
```
squashed and lightly re-ordered

instagibbs committed:
Add release note for size 2 package rbf

theStack reviewed (commented):
Spent most time reviewing the core commit 612847ae1a55e92bb7732905a026be96a436372a so far, which looks correct to me (though I'm not super-familiar with the details of the feerate diagram checks that are used in this PR for the first time in production code). Still planning on looking deeper at the tests and do another full review round. Left some non-blocking nits on the way.

instagibbs commented on src/validation.cpp:
```diff
@@ -1141,7 +1140,81 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    }
+
+    // Check if it's economically rational to mine this package rather than the ones it replaces.
+    // This takes the place of ReplacementChecks()'s PaysMoreForConflicts() in the package RBF setting.
```
done

instagibbs commented on src/validation.cpp:
```diff
@@ -1141,7 +1140,81 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    // to be only paying anti-DoS fees
+    const CFeeRate parent_feerate(parent_ws.m_modified_fees, parent_ws.m_vsize);
+    const CFeeRate package_feerate(m_subpackage.m_total_modified_fees, m_subpackage.m_total_vsize);
+    if (parent_feerate >= package_feerate) {
```
done

instagibbs commented on src/validation.cpp:
```diff
@@ -1141,7 +1140,81 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
+    }
+
+    LogPrint(BCLog::TXPACKAGES, "package RBF checks passed: parent %s (wtxid=%s), child %s (wtxid=%s)\n",
+        txns.front()->GetHash().ToString(), txns.front()->GetWitnessHash().ToString(),
+        txns.back()->GetHash().ToString(), txns.back()->GetWitnessHash().ToString());
```
added at beginning of function

instagibbs commented on src/test/txpackage_tests.cpp:
```diff
@@ -938,4 +939,149 @@ BOOST_FIXTURE_TEST_CASE(package_cpfp_tests, TestChain100Setup)
+    LOCK(::cs_main);
+    size_t expected_pool_size = m_node.mempool->size();
+    CKey child_key;
+    child_key.MakeNewKey(true);
```
done

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,589 @@
+
+        # Make singleton tx to conflict with in next batch
+        singleton_coin = self.coins[-1]
+        del self.coins[-1]
```
done in this and one more location

instagibbs commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,28 @@ The following rules are enforced for all packages:
+
+   - All conflicting clusters(connected components of mempool transactions) must be clusters of up to size 2.
+
+   - No more than MAX_REPLACEMENT_CANDIDATES transactions can be replaced.
```
done

instagibbs commented:
@theStack part of splitting out the diagram stuff to a prep PR is that you can just assume it means "economically rational to replace" :+1:

glozow reviewed (commented):
reACK 94ed4fbf8e via range-diff

ismaelsadeeq reviewed (approved):
re-ACK 94ed4fbf8e1a396c650b5134d396d6c0be35ce10

theStack reviewed (approved):
Code-review ACK 94ed4fbf8e1a396c650b5134d396d6c0be35ce10

murchandamus reviewed (commented):
utACK 94ed4fbf8e1a396c650b5134d396d6c0be35ce10

murchandamus commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,29 @@ The following rules are enforced for all packages:
 * Packages cannot have conflicting transactions, i.e. no two transactions in a package can spend
    the same inputs. Packages cannot have duplicate transactions. (#20833)
```
I think this comment is now also outdated?

instagibbs commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,29 @@ The following rules are enforced for all packages:
+
+   - Packages are 1-parent-1-child, with no in-mempool ancestors of the package.
+
+   - All conflicting clusters(connected components of mempool transactions) must be clusters of up to size 2.
```
done in follow-up

instagibbs commented on doc/policy/packages.md:
```diff
@@ -36,10 +36,29 @@ The following rules are enforced for all packages:
+* Only limited package replacements are currently considered. (#28984)
 
-   - Package RBF may be enabled in the future.
+   - All direct conflicts must signal replacement (or have `-mempoolfullrbf=1` set).
```
done in follow-up

instagibbs commented on src/test/util/txmempool.cpp:
```diff
@@ -68,6 +69,28 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
+        // Replacements can't happen for subpackages larger than 2
+        if (!atmp_result.m_replaced_transactions.empty() &&
+            atmp_result.m_wtxids_fee_calculations.has_value() && atmp_result.m_wtxids_fee_calculations.value().size() > 2) {
+             return strprintf("tx %s was part of a too-large package RBF subpackage",
+                                wtxid.ToString());
```
done in follow-up

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,587 @@
+        assert_equal(f"package RBF failed: insufficient anti-DoS fees, rejecting replacement {package_txns3[1].rehash()}, not enough additional fees to relay; 0.00 < 0.00000{sum([tx.get_vsize() for tx in package_txns3])}", pkg_results3["package_msg"])
+
+        self.assert_mempool_contents(expected=package_txns1)
+        self.generate(node, 1)
```
wrote a better boundary test in this vein in follow-up, thanks. You actually need incremental added, not 1 sat.

instagibbs commented on test/functional/mempool_package_rbf.py:
```diff
@@ -0,0 +1,587 @@
+        assert 'package RBF failed: package feerate is less than parent feerate' in pkg_results5["package_msg"]
+
+        self.assert_mempool_contents(expected=package_txns4)
+        self.generate(node, 1)
```
took as inspiration to make an actual boundary-condition test in follow-up, thanks

instagibbs commented:
opened followup at https://github.com/bitcoin/bitcoin/pull/30295

instagibbs commented on src/test/util/txmempool.cpp:
```diff
@@ -68,6 +69,28 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
             return strprintf("tx %s unexpectedly failed: %s", wtxid.ToString(), atmp_result.m_state.ToString());
         }
 
+        // Each subpackage is allowed MAX_REPLACEMENT_CANDIDATES replacements (only checking individually here)
+        if (atmp_result.m_replaced_transactions.size() > MAX_REPLACEMENT_CANDIDATES) {
+            return strprintf("tx %s result replaced too many transactions",
+                                wtxid.ToString());
+        }
```
done in follow-up

glozow replied:
I don't agree with this suggestion - we are checking the result of a transaction that *was* accepted here. So "would replace" is incorrect and "replaced" is correct.

achow101 commented:
ACK 94ed4fbf8e1a396c650b5134d396d6c0be35ce10

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/264.
